### ISSUE TITLE
Migration and testing for Grouper v2.5

### ISSUE THREAD

**Issue by astrochun** [UAL-RE/ReQUIAM#116]
**Summary**
UITS is migrating from v2.2 to v2.5 of Grouper. We will need to perform checks using the test instance before it goes live.

Deadline: Comments/issues to Mark Fischer by 03/16/2021
Upgrade is schedule to occur on 04/05/2021.

**Objectives**
Ensure that migration to v2.5 will work

**Proposal**
- [x] Develop a feature branch with new endpoint (EDIT: Easier to create a new config file for this to test than to have new flags that will most likely be deprecated ~testing flags to work with new endpoint~).
- [x] Run tests:
    - [x] Run `script_run` with `--portal`
    - [x] Run `script_run` with `--test` and `--sync` [FAILED - Error provided below]
    - [x] Run `script_run` with `--test_reverse` and `--sync`
    - [x] Run `user_update` script on a test account
      - [x] `--portal <portal>`
      - [x] `--portal root` [This can't be done since the member does not exist in a given portal with no full end-to-end testing]
    - [x] Run `add_grouper_groups` without a new group on dry-run:
        - [x] On `figtest` stem
            - [x] Execute with `--main_themes` [EDIT: See [comment](https://github.com/ualibraries/ReQUIAM/issues/116#issuecomment-788997947) ~some groups were missing~]
            - [x] Execute with `--sub_portals` [EDIT: See [comment](https://github.com/ualibraries/ReQUIAM/issues/116#issuecomment-788997947) ~some groups were missing~]
            - [x] Execute with `--quota` [EDIT: See [comment](https://github.com/ualibraries/ReQUIAM/issues/116#issuecomment-788997947) ~some groups were missing~]
        - [x] On `figshare` stem (production)
            - [x] Execute with `--main_themes`
            - [x] Execute with `--sub_portals`
            - [x] Execute with `--quota`
    - [x] Test group creation via the `grouper.create_groups` function
    - [x] Run `add_grouper_groups` with a new test group on `figtest` stem

**Deployment**
Change to connection strings is needed when v2.5 will be deployed by UITS. In the mean time, we wait and continue to use existing configuration.

*Testing notes*

The new testing endpoint is:https://grouper25.iam.arizona.edu/grouper-ws/  (Not https://grouper25.test.iam.arizona.edu/grouper-ws)   

grouper_host = grouper25.iam.arizona.edu
grouper_base_path = grouper-ws/servicesRest/json/v2_5_001

*Additional notes*

This should be performed after #87 is completed given the significant amount of possible refactoring involved.

**Implemented in**: `feature/116_grouper25`

**Comment by astrochun** [comment]
Testing:
1. Attempted to run `script_run` with `--test` and `--sync` resulted in the following error:

``` python
11:45:39 -     INFO: ldap and grouper have 0 members in common
11:45:39 -     INFO: synchronization will drop 0 entries from grouper group
11:45:39 -     INFO: synchronization will add 1 entries to grouper group
11:45:39 -     INFO: adds  : ['***REDACTED***']
11:45:39 -     INFO: synchronizing ...
11:45:39 -     INFO: synchronizing ldap query results to arizona.edu:dept:LBRY:figshare:test
11:45:39 -     INFO: batch size = 100, batch timeout = 60 seconds, batch delay = 0 seconds
11:45:39 -     INFO: processing drops:
11:45:39 -     INFO: processing adds:
11:45:39 -  WARNING: problem running batch add, result code = PROBLEM_WITH_ASSIGNMENT
11:45:39 -     INFO: TEST_SYNC : Total time:  0 hours   0 minutes   1.63 seconds
11:45:39 -     INFO: Total time:  0 hours   0 minutes   3.27 seconds
```

Note that when I run it with the v2.2 config, it actually says the user is in the group so the change was implemented but `grouper25` UI does not show the update.

**Comment by astrochun** [comment]
The issue stems not from our end but on `grouper25` test instance. The test accounts was not enabled at that time. Thus, it resulted in an error with the following message: "subject not found: ***REDACTED***"

**Comment by astrochun** [comment]
I learned that there is a test EDS instance with the URL of `eds.test.iam.arizona.edu` that is synchronize with the Grouper v2.5 test instance. However, when I attempted to run `user_update`, the script warned that the user was not available:
```
09:27:01 -  WARNING: No ismembersof attributes
09:27:01 -     INFO: --netid <USER> -test --active_remove --portal root --quota root 
09:27:01 -     INFO: None of the supplied users are active
09:27:01 -     INFO: List of users that are NOT active : ['<USER>']
09:27:01 -     INFO: Main stem: arizona.edu:dept:LBRY:figshare
09:27:01 -     INFO: STAGE: ACTIVE
09:27:01 -     INFO: Creating figtest:group_active group
09:27:01 -     INFO: dry run, not creating active_group group
09:27:01 -     INFO: STAGE: PORTAL
```

This is even the case with my netid, which I also have direct membership to the `active` group.

**Comment by astrochun** [comment]
When testing `add_grouper_groups` there are missing groups:

``` python
10:39:50 -     INFO: *************************************
10:39:51 -     INFO: Started add_grouper_group script ... 
10:40:22 -     INFO:               config = config/figshare_grouper25.ini
10:40:22 -     INFO:      persistent_path = ***REDACTED***
10:40:22 -     INFO:         grouper_host = grouper25.test.iam.arizona.edu
10:40:22 -     INFO:    grouper_base_path = grouper-ws/servicesRest/json/v2_5_001
10:40:22 -     INFO:         grouper_user = (set)
10:40:22 -     INFO:     grouper_password = (set)
10:40:22 -     INFO:          main_themes = True
10:40:22 -     INFO:          sub_portals = False
10:40:22 -     INFO:                quota = False
10:40:22 -     INFO:           production = False
10:40:22 -     INFO:                  add = False
10:40:22 -     INFO:                debug = False
10:40:22 -     INFO: Reading Main Theme CSV into DataFrame ...
10:40:25 -     INFO: Excluded main themes: restricted_publish
10:40:25 -     INFO: Total number of main themes: 10
10:40:25 -     INFO: List of main themes: arts_design, business_econ, education, humanities, law, med_health, sci_math, social_sci, tech_eng, covid19
10:40:26 -     INFO: Group exists : arts_design
10:40:27 -     INFO: dry run, not performing privilege add
10:40:27 -     INFO: Group does not exist : business_econ
10:40:27 -     INFO: dry run, not performing group add
10:40:27 -     INFO: dry run, not performing privilege add
10:40:28 -     INFO: Group does not exist : education
10:40:28 -     INFO: dry run, not performing group add
10:40:28 -     INFO: dry run, not performing privilege add
10:40:28 -     INFO: Group does not exist : humanities
10:40:28 -     INFO: dry run, not performing group add
10:40:28 -     INFO: dry run, not performing privilege add
10:40:29 -     INFO: Group does not exist : law
10:40:29 -     INFO: dry run, not performing group add
10:40:29 -     INFO: dry run, not performing privilege add
10:40:30 -     INFO: Group does not exist : med_health
10:40:30 -     INFO: dry run, not performing group add
10:40:30 -     INFO: dry run, not performing privilege add
10:40:30 -     INFO: Group exists : sci_math
10:40:30 -     INFO: dry run, not performing privilege add
10:40:31 -     INFO: Group exists : social_sci
10:40:31 -     INFO: dry run, not performing privilege add
10:40:31 -     INFO: Group exists : tech_eng
10:40:31 -     INFO: dry run, not performing privilege add
10:40:32 -     INFO: Group exists : covid19
10:40:32 -     INFO: dry run, not performing privilege add
10:40:32 -     INFO: MAIN PORTAL : Total time:  0 hours   0 minutes   9.58 seconds
10:40:32 -     INFO: Total time:  0 hours   0 minutes   41.48 seconds
10:40:32 -     INFO: *************************************
10:40:32 -     INFO: Exit 0
```

Additional details provided in: add_grouper_groups.2021-02-03.log

**Comment by astrochun** [comment]
The missing groups (when I ran `add_grouper_groups` was due to a sync where groups were not created if they were empty. This was resolved by UITS Tim Darby.

**Comment by astrochun** [comment]
With synchronization from grouper2.2 -> grouper2.5, the `pgrps` stem should be updated for testing purposes.

**Comment by astrochun** [comment]
We learned from UITS that test EDS is updated with Grouper 2.5 updates. I've gone ahead and done a manual sync. Here's the result from a "dry run":
```
15:39:32     INFO -           script_run                       <module> : SUMMARY DATA
|                  |   num_EDS |   num_Grouper |   adds |   drops |   total |
|:-----------------|----------:|--------------:|-------:|--------:|--------:|
| fine_arts        |       264 |           264 |      0 |       0 |       0 |
| performing_arts  |       441 |           441 |      0 |       0 |       0 |
| architecture     |       190 |           189 |      1 |       0 |       1 |
| arts_design      |        37 |            37 |      0 |       0 |       0 |
| business_econ    |       429 |           429 |      0 |       0 |       0 |
| management       |       245 |           245 |      0 |       0 |       0 |
| economics        |       103 |           103 |      0 |       0 |       0 |
| education        |      1377 |          1376 |      1 |       0 |       1 |
| english          |       291 |           291 |      0 |       0 |       0 |
| lang_culture     |       375 |           375 |      0 |       0 |       0 |
| humanities       |       247 |           246 |      1 |       0 |       1 |
| law              |       397 |           397 |      0 |       0 |       0 |
| nursing          |       337 |           337 |      0 |       0 |       0 |
| med_health       |      2020 |          2021 |      1 |       2 |       3 |
| clinical         |       862 |           861 |      1 |       0 |       1 |
| ped_reprod       |       203 |           202 |      1 |       0 |       1 |
| neurology        |       112 |           112 |      0 |       0 |       0 |
| oncology         |       276 |           277 |      0 |       1 |       1 |
| pharmacology     |       732 |           732 |      0 |       0 |       0 |
| physiology       |       142 |           142 |      1 |       1 |       2 |
| public_health    |       490 |           489 |      1 |       0 |       1 |
| astro            |       623 |           623 |      0 |       0 |       0 |
| cognitive_sci    |       353 |           353 |      0 |       0 |       0 |
| life_sci         |      1764 |          1762 |      2 |       0 |       2 |
| sci_math         |      1942 |          1943 |      1 |       2 |       3 |
| earth_sci        |      1057 |          1057 |      0 |       0 |       0 |
| physics          |       205 |           205 |      0 |       0 |       0 |
| lpl              |       322 |           322 |      0 |       0 |       0 |
| agriculture      |       135 |           135 |      0 |       0 |       0 |
| anthropology     |       211 |           211 |      0 |       0 |       0 |
| social_sci       |       870 |           868 |      2 |       0 |       2 |
| cultural_studies |       167 |           166 |      1 |       0 |       1 |
| history          |        77 |            77 |      0 |       0 |       0 |
| journalism       |       257 |           257 |      0 |       0 |       0 |
| engineering      |      1337 |          1335 |      2 |       0 |       2 |
| technology       |       158 |           158 |      0 |       0 |       0 |
| 536870912        |     12914 |         12914 |      0 |       0 |       0 |
| 2147483648       |     25481 |         25481 |      0 |       0 |       0 |
```

Here are the results for v2.2 (ran right afterwards)
```
15:50:37 -     INFO: SUMMARY DATA
|                  |   num_EDS |   num_Grouper |   adds |   drops |   total |
|:-----------------|----------:|--------------:|-------:|--------:|--------:|
| fine_arts        |       265 |           265 |      0 |       0 |       0 |
| performing_arts  |       441 |           441 |      0 |       0 |       0 |
| architecture     |       189 |           188 |      1 |       0 |       1 |
| arts_design      |        38 |            38 |      0 |       0 |       0 |
| business_econ    |       429 |           429 |      0 |       0 |       0 |
| management       |       244 |           244 |      0 |       0 |       0 |
| economics        |       103 |           103 |      0 |       0 |       0 |
| education        |      1391 |          1391 |      1 |       1 |       2 |
| english          |       291 |           291 |      0 |       0 |       0 |
| lang_culture     |       383 |           383 |      0 |       0 |       0 |
| humanities       |       246 |           245 |      1 |       0 |       1 |
| law              |       396 |           396 |      0 |       0 |       0 |
| nursing          |       331 |           332 |      0 |       1 |       1 |
| med_health       |      2044 |          2045 |      1 |       2 |       3 |
| clinical         |       862 |           861 |      1 |       0 |       1 |
| ped_reprod       |       201 |           201 |      1 |       1 |       2 |
| neurology        |       112 |           112 |      0 |       0 |       0 |
| oncology         |       277 |           278 |      0 |       1 |       1 |
| pharmacology     |       736 |           735 |      1 |       0 |       1 |
| physiology       |       142 |           142 |      1 |       1 |       2 |
| public_health    |       500 |           499 |      1 |       0 |       1 |
| astro            |       616 |           616 |      0 |       0 |       0 |
| cognitive_sci    |       353 |           353 |      0 |       0 |       0 |
| life_sci         |      1783 |          1782 |      2 |       1 |       3 |
| sci_math         |      1930 |          1932 |      1 |       3 |       4 |
| earth_sci        |      1062 |          1062 |      0 |       0 |       0 |
| physics          |       205 |           205 |      0 |       0 |       0 |
| lpl              |       324 |           324 |      0 |       0 |       0 |
| agriculture      |       136 |           136 |      0 |       0 |       0 |
| anthropology     |       208 |           208 |      0 |       0 |       0 |
| social_sci       |       876 |           874 |      2 |       0 |       2 |
| cultural_studies |       167 |           166 |      1 |       0 |       1 |
| history          |        77 |            77 |      0 |       0 |       0 |
| journalism       |       256 |           256 |      0 |       0 |       0 |
| engineering      |      1349 |          1347 |      2 |       0 |       2 |
| technology       |       111 |           111 |      0 |       0 |       0 |
| 536870912        |     14784 |         14786 |      3 |       5 |       8 |
| 2147483648       |     24113 |         24124 |      1 |      12 |      13 |
```

Note: There are differences in numbers.

**Comment by astrochun** [comment]
Note that this issue is not closed. The changes were merged into `develop` branch earlier. It is still on hold/blocked due to database changes by UITS for EDS.

Latest update from UITS (Tim Darby):
> I'm in the process of testing the move to production and we'll soon put out a date. We want to build in enough time for another round of testing by campus users.

**Comment by astrochun** [comment]
Here's the latest run on 2.2:

```
|                  |   num_EDS |   num_Grouper |   adds |   drops |   total |
|:-----------------|----------:|--------------:|-------:|--------:|--------:|
| fine_arts        |       239 |           239 |      0 |       0 |       0 |
| performing_arts  |       382 |           383 |      0 |       1 |       1 |
| architecture     |       154 |           154 |      0 |       0 |       0 |
| arts_design      |        40 |            40 |      0 |       0 |       0 |
| business_econ    |       401 |           401 |      0 |       0 |       0 |
| management       |       217 |           217 |      0 |       0 |       0 |
| economics        |        98 |            98 |      0 |       0 |       0 |
| education        |      1278 |          1278 |      0 |       0 |       0 |
| english          |       265 |           265 |      0 |       0 |       0 |
| lang_culture     |       347 |           347 |      0 |       0 |       0 |
| humanities       |       242 |           242 |      0 |       0 |       0 |
| law              |       381 |           381 |      0 |       0 |       0 |
| nursing          |       346 |           346 |      0 |       0 |       0 |
| med_health       |      1903 |          1904 |      0 |       1 |       1 |
| clinical         |       886 |           886 |      0 |       0 |       0 |
| ped_reprod       |       202 |           202 |      0 |       0 |       0 |
| neurology        |       108 |           108 |      0 |       0 |       0 |
| oncology         |       282 |           281 |      1 |       0 |       1 |
| pharmacology     |       716 |           716 |      0 |       0 |       0 |
| physiology       |       128 |           128 |      0 |       0 |       0 |
| public_health    |       472 |           469 |      3 |       0 |       3 |
| astro            |       621 |           621 |      0 |       0 |       0 |
| cognitive_sci    |       355 |           355 |      0 |       0 |       0 |
| life_sci         |      1763 |          1761 |      2 |       0 |       2 |
| sci_math         |      1785 |          1784 |      1 |       0 |       1 |
| earth_sci        |      1025 |          1026 |      0 |       1 |       1 |
| physics          |       199 |           199 |      0 |       0 |       0 |
| lpl              |       306 |           306 |      0 |       0 |       0 |
| agriculture      |       134 |           134 |      0 |       0 |       0 |
| anthropology     |       206 |           206 |      0 |       0 |       0 |
| social_sci       |       840 |           840 |      0 |       0 |       0 |
| cultural_studies |       160 |           160 |      0 |       0 |       0 |
| history          |        74 |            74 |      0 |       0 |       0 |
| journalism       |       240 |           240 |      0 |       0 |       0 |
| engineering      |      1205 |          1205 |      0 |       0 |       0 |
| technology       |       109 |           109 |      0 |       0 |       0 |
| libraries        |       279 |           279 |      0 |       0 |       0 |
| 536870912        |     15650 |         15650 |      0 |       0 |       0 |
| 2147483648       |     23426 |         23426 |      0 |       0 |       0 |
```

Here are the numbers from Grouper 2.5 after database migration/upgrade.
Note: EDS production, grouper25 test:

```
|                  |   num_EDS |   num_Grouper |   adds |   drops |   total |
|:-----------------|----------:|--------------:|-------:|--------:|--------:|
| fine_arts        |       239 |           239 |      3 |       3 |       6 |
| performing_arts  |       382 |           382 |      3 |       3 |       6 |
| architecture     |       154 |           153 |      1 |       0 |       1 |
| arts_design      |        40 |            37 |      3 |       0 |       3 |
| business_econ    |       401 |           397 |      4 |       0 |       4 |
| management       |       217 |           215 |      3 |       1 |       4 |
| economics        |        98 |            96 |      2 |       0 |       2 |
| education        |      1278 |          1267 |     19 |       8 |      27 |
| english          |       265 |           266 |      1 |       2 |       3 |
| lang_culture     |       347 |           343 |      5 |       1 |       6 |
| humanities       |       242 |           240 |      2 |       0 |       2 |
| law              |       381 |           380 |      3 |       2 |       5 |
| nursing          |       346 |           337 |      9 |       0 |       9 |
| med_health       |      1903 |          1928 |     20 |      45 |      65 |
| clinical         |       886 |           880 |      9 |       3 |      12 |
| ped_reprod       |       202 |           201 |      2 |       1 |       3 |
| neurology        |       108 |           109 |      0 |       1 |       1 |
| oncology         |       282 |           276 |     12 |       6 |      18 |
| pharmacology     |       716 |           708 |     12 |       4 |      16 |
| physiology       |       128 |           127 |      1 |       0 |       1 |
| public_health    |       472 |           472 |      7 |       7 |      14 |
| astro            |       621 |           621 |      5 |       5 |      10 |
| cognitive_sci    |       355 |           350 |      6 |       1 |       7 |
| life_sci         |      1763 |          1744 |     27 |       8 |      35 |
| sci_math         |      1785 |          1780 |     14 |       9 |      23 |
| earth_sci        |      1025 |          1021 |     12 |       8 |      20 |
| physics          |       199 |           200 |      1 |       2 |       3 |
| lpl              |       306 |           306 |      2 |       2 |       4 |
| agriculture      |       134 |           132 |      2 |       0 |       2 |
| anthropology     |       206 |           204 |      2 |       0 |       2 |
| social_sci       |       840 |           833 |     10 |       3 |      13 |
| cultural_studies |       160 |           160 |      1 |       1 |       2 |
| history          |        74 |            74 |      0 |       0 |       0 |
| journalism       |       240 |           239 |      3 |       2 |       5 |
| engineering      |      1205 |          1197 |     12 |       4 |      16 |
| technology       |       109 |           112 |      0 |       3 |       3 |
| 536870912        |     15650 |         15466 |    266 |      82 |     348 |
| 2147483648       |     23426 |         23186 |    406 |     166 |     572 |
```

**Comment by astrochun** [comment]
There are two issues:
1. There are differences
2. For some reason, the new libraries group is missing in grouper25. However, I checked Grouper UI and it's present
```
09:21:13     INFO -           script_run                       <module> : Working on University Libraries (libraries) portal
09:21:14  WARNING -           script_run                       <module> : !!! Grouper portal NOT found : libraries !!!
```

Actually upon investigation, this is due to using the wrong Grouper 2.5 testing instance. Previously we had grouper25.test.iam.arizona.edu. It is now grouper25.iam.arizona.edu. This is fixed in b78ce1f. Here are revised numbers:

```
13:31:01 -     INFO: SUMMARY DATA
|                  |   num_EDS |   num_Grouper |   adds |   drops |   total |
|:-----------------|----------:|--------------:|-------:|--------:|--------:|
| fine_arts        |       240 |           239 |      2 |       1 |       3 |
| performing_arts  |       382 |           382 |      1 |       1 |       2 |
| architecture     |       154 |           154 |      0 |       0 |       0 |
| arts_design      |        39 |            40 |      0 |       1 |       1 |
| business_econ    |       401 |           400 |      1 |       0 |       1 |
| management       |       217 |           217 |      0 |       0 |       0 |
| economics        |        98 |            97 |      1 |       0 |       1 |
| education        |      1278 |          1269 |     10 |       1 |      11 |
| english          |       265 |           264 |      1 |       0 |       1 |
| lang_culture     |       347 |           347 |      0 |       0 |       0 |
| humanities       |       242 |           242 |      0 |       0 |       0 |
| law              |       381 |           380 |      1 |       0 |       1 |
| nursing          |       346 |           343 |      3 |       0 |       3 |
| med_health       |      1902 |          1905 |      1 |       4 |       5 |
| clinical         |       886 |           887 |      0 |       1 |       1 |
| ped_reprod       |       202 |           202 |      0 |       0 |       0 |
| neurology        |       108 |           108 |      0 |       0 |       0 |
| oncology         |       282 |           282 |      4 |       4 |       8 |
| pharmacology     |       716 |           715 |      1 |       0 |       1 |
| physiology       |       128 |           127 |      1 |       0 |       1 |
| public_health    |       472 |           470 |      3 |       1 |       4 |
| astro            |       621 |           619 |      2 |       0 |       2 |
| cognitive_sci    |       355 |           354 |      1 |       0 |       1 |
| life_sci         |      1763 |          1759 |      6 |       2 |       8 |
| sci_math         |      1785 |          1784 |      4 |       3 |       7 |
| earth_sci        |      1031 |          1027 |      7 |       3 |      10 |
| physics          |       198 |           199 |      0 |       1 |       1 |
| lpl              |       307 |           306 |      2 |       1 |       3 |
| agriculture      |       132 |           134 |      0 |       2 |       2 |
| anthropology     |       205 |           206 |      0 |       1 |       1 |
| social_sci       |       839 |           841 |      0 |       2 |       2 |
| cultural_studies |       160 |           160 |      0 |       0 |       0 |
| history          |        74 |            74 |      0 |       0 |       0 |
| journalism       |       240 |           239 |      1 |       0 |       1 |
| engineering      |      1205 |          1201 |      6 |       2 |       8 |
| technology       |       110 |           109 |      1 |       0 |       1 |
| libraries        |       280 |           277 |      3 |       0 |       3 |
| 536870912        |     15653 |         15640 |     35 |      22 |      57 |
| 2147483648       |     23432 |         23411 |     49 |      28 |      77 |
```

Note that the above was synced by 07/28 @ 1:31PM.

**Comment by yhan818** [comment]
**Tested grouper 2.5 settings:**

grouper_host = grouper25.iam.arizona.edu
grouper_base_path = grouper-ws/servicesRest/json/v2_5_001

**Original grouper 2.2 settings:**
grouper_host = grouper.iam.arizona.edu
grouper_base_path = grouper-ws/servicesRest/json/v2_2_001

**Comment by yhan818** [comment]
Dry runs of the 2.5 vs. 2.2 within 10 min show a few differences: 
![Screenshot from 2021-08-03 15-54-39](https://user-images.githubusercontent.com/8713681/128096820-740ac6c0-769b-4306-adb3-8e019513beeb.png)

**Comment by yhan818** [comment]
Dry runs again after UITS performed sync on grouper this morning. The results are the same at Aug 4 at 12 PM

**Comment by yhan818** [comment]
Grouper v2.5 was updated by UITS on Sep 4, 2021. With a few things updated, everything runs normal.
